programs/gateway/src/lib.rs

```diff
@@ -396,4 +396,11 @@ pub mod gateway {
             nonce,
         )
     }
+
+    #[cfg(feature = "upgrade")]
```

Comment:
this removes the necessity of maintaining two separate programs , its cleaner imo, but I can revert to the old strategy if needed